File: src/lib/faUrls.ts
const BASE_URL = 'https://www.furaffinity.net';

export const FAUrls = {
  home: BASE_URL,
  login: `${BASE_URL}/login`,
  register: `${BASE_URL}/register`,
  submissions: `${BASE_URL}/msg/submissions/new@72`,
  submissionsFrom: (sid: number) => `${BASE_URL}/msg/submissions/new~${sid}@72`,
  browse: (filter?: string, page?: number) => {
    const f = filter && filter !== 'all' ? `?type=${filterMap[filter as keyof typeof filterMap] || ''}` : '';
    const p = page && page > 1 ? `&page=${page}` : '';
    return `${BASE_URL}/browse/all/${f}${p}`;
  },
  viewSubmission: (id: string) => `${BASE_URL}/view/${id}/`,
  search: (query: string, page?: number) =>
    `${BASE_URL}/search/?q=${encodeURIComponent(query)}${page && page > 1 ? `&page=${page}` : ''}`,
  notifications: `${BASE_URL}/msg/others/`,
  user: (username: string) => `${BASE_URL}/user/${username}/`,
  avatar: (username: string) => `https://a.furaffinity.net/${username}.gif`,
  gallery: (username: string) => `${BASE_URL}/gallery/${username}/`,
  favorites: (username: string) => `${BASE_URL}/favorites/${username}/`,
  journals: (username: string) => `${BASE_URL}/journals/${username}/`,
  journal: (id: string) => `${BASE_URL}/journal/${id}/`,
  watchlist: (username: string, direction: 'to' | 'by', page?: number) =>
    `${BASE_URL}/watchlist/${direction}/${username}/${page && page > 1 ? `?page=${page}` : ''}`,
  notesInbox: `${BASE_URL}/controls/switchbox/inbox/`,
  notesSent: `${BASE_URL}/controls/switchbox/sent/`,
  newNote: (username: string) => `${BASE_URL}/newpm/${username}/`,
};

const filterMap = {
  all: '',
  digital: '1',
  traditional: '2',
  writing: '3',
};

File: furclient/lib/services/fa_urls.dart
class FAUrls {
  static const String baseUrl = 'https://www.furaffinity.net';

  static String get home => baseUrl;
  static String get login => '$baseUrl/login';
  static String get register => '$baseUrl/register';
  static String get submissions => '$baseUrl/msg/submissions/new@72';
  static String submissionsFrom(int sid) =>
      '$baseUrl/msg/submissions/new~$sid@72';

  static String browse({String filter = 'all', int? page}) {
    final params = <String>[];
    if (filter != 'all' && filterMap.containsKey(filter)) {
      params.add('type=${filterMap[filter]}');
    }
    if (page != null && page > 1) {
      params.add('page=$page');
    }
    final query = params.isNotEmpty ? '?${params.join('&')}' : '';
    return '$baseUrl/browse/$query';
  }

  static String viewSubmission(String id) => '$baseUrl/view/$id/';
  static String search(String query, {int page = 1}) {
    final q = Uri.encodeComponent(query);
    final p = page > 1 ? '&page=$page' : '';
    return '$baseUrl/search/?q=$q$p';
  }

  static String get notifications => '$baseUrl/msg/others/';
  static String user(String username) => '$baseUrl/user/$username/';
  static String avatar(String username) =>
      'https://a.furaffinity.net/$username.gif';
  static String gallery(String username) => '$baseUrl/gallery/$username/';
  static String favorites(String username) =>
      '$baseUrl/favorites/$username/';
  static String journals(String username) => '$baseUrl/journals/$username/';
  static String journal(String id) => '$baseUrl/journal/$id/';

  static String watchlist(String username, String direction, {int? page}) {
    final p = (page != null && page > 1) ? '?page=$page' : '';
    return '$baseUrl/watchlist/$direction/$username/$p';
  }

  static String get notesInbox =>
      '$baseUrl/controls/switchbox/inbox/';
  static String get notesSent =>
      '$baseUrl/controls/switchbox/sent/';
  static String newNote(String username) => '$baseUrl/newpm/$username/';

  static const Map<String, String> filterMap = {
    'all': '',
    'digital': '1',
    'traditional': '2',
    'writing': '3',
  };
}
